Update ingestion spec doc (#13329)
apache/druid: docs/ingestion/ingestion-spec.md

|-----|-----------|-------|
|bitmap|Compression format for bitmap indexes. Should be a JSON object with `type` set to `roaring` or `concise`. For type `roaring`, the boolean property `compressRunOnSerialization` (defaults to true) controls whether or not run-length encoding will be used when it is determined to be more space-efficient.|`{"type": "roaring"}`|
|dimensionCompression|Compression format for dimension columns. Options are `lz4`, `lzf`, `zstd`, or `uncompressed`.|`lz4`|

> |stringDictionaryEncoding|Encoding format for STRING-typed column value dictionaries. The default setting `utf8` suits most use cases.<br>Example to enable front coding: `{"type":"frontCoded", "bucketSize": 4}`<br>`bucketSize` is the number of values to place in a bucket to perform delta encoding. Must be a power of 2, maximum is 128. Defaults to 4.<br>See [Front coding](#front-coding) for more information.|`{"type":"utf8"}`|

Reviewer: hmm, thinking a bit more about this, I wonder if the additional details here would be enough if we also added the bit about being new to Druid 25.0, and instead drop all of the extra sections dedicated to `stringDictionaryEncoding`.

My reasoning is that these `indexSpec` settings provide direct control over low level segment storage details. They are primarily intended for experimentation by _the Druid developers_ to assist in determining the best default values for these settings, though i imagine advanced cluster operators also use them in the process of trying to maximally optimize a cluster.

We have to have all of this text to describe implementation details of the `frontCoded` option so that the operator can begin to understand how to correctly set the `bucketSize` parameter, and the implications of different sizes also require a decent bit of understanding of both the data and the way Druid works internally.

Did we add enough information to make an informed decision? I'm not sure if I'm a good judge of that 😅 If it is actually enough to understand reasonably well I can be convinced to leave all of these details.

Author: Updated.

> |stringDictionaryEncoding|Encoding format for STRING-typed column value dictionaries. The default setting `utf8` suits most use cases.<br>Example to enable front coding: `{"type":"frontCoded", "bucketSize": 4}`<br>`bucketSize` is the number of values to place in a bucket to perform delta encoding. Must be a power of 2, maximum is 128. Defaults to 4.<br>See [Front coding](#front-coding) for more information.|`{"type":"utf8"}`|

Reviewer: suggest clarifying that applies to `STRING` and `COMPLEX<json>` columns
```suggestion
|stringDictionaryEncoding|Encoding format for STRING value dictionaries used by `STRING` and `COMPLEX<json>` columns. <br>Example to enable front coding: `{"type":"frontCoded", "bucketSize": 4}`<br>`bucketSize` is the number of values to place in a bucket to perform delta encoding. Must be a power of 2, maximum is 128. Defaults to 4.<br>See [Front coding](#front-coding) for more information.|`{"type":"utf8"}`|
```

Author: Updated.

##### Front coding

By default, Druid stores values in STRING-typed columns as uncompressed UTF-8 encoded bytes.

Starting in version 25.0, Druid can store STRING columns using an incremental encoding strategy called front coding. This allows Druid to create smaller UTF-8 encoded segments with very little performance cost.

If you enable front coding, Druid divides the column values into buckets, storing the first value in each bucket as it is. Druid stores subsequent values in the bucket using a number representing the length of the prefix and the remainder of the value. This technique prevents Druid from storing duplicated prefixes.

If you set `bucketSize` to a larger number than the default, larger buckets allow columns with a high degree of overlap to produce smaller segments. This change causes a slight cost to read and search performance which scales with bucket size.

Example `indexSpec` snippet with front coding enabled:

```plaintext
"indexSpec": {
    "bitmap": { "type": "roaring" },
    "dimensionCompression": "lz4",
    "metricCompression": "lz4",
    "jsonCompression": "lz4",
    "longEncoding": "auto",
    "stringDictionaryEncoding": {
      "type": "utf8"
    }

>   }

Author: I think the additional detail in the table means that we can omit this.
